### PR TITLE
[Feature] Support TIMMBackbone

### PR DESCRIPTION
**Pull request by shinya7y**
## Motivation

timm (https://github.com/rwightman/pytorch-image-models) contains a lot of backbones, which are not available in this repository. It would be convenient to add examples to use timm backbones.

Related to the features of mmcls and mmseg.
https://github.com/open-mmlab/mmclassification/pull/427
https://github.com/open-mmlab/mmsegmentation/pull/998

## Modification

This PR adds example configs and a document to use timm backbones via the `TIMMBackbone` wrapper in mmcls.

## Use cases (Optional)

Please see config files.

We can use ConvNeXt with timm>=0.5.4. However, other efforts are needed to support the settings of the [official ConvNeXt](https://github.com/facebookresearch/ConvNeXt) (e.g., `LearningRateDecayOptimizerConstructor`). Related to https://github.com/open-mmlab/mmsegmentation/pull/1216

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -76,6 +76,7 @@ jobs:
       - run:
           name: Run unittests
           command: |
+            pip install timm
```
Comment (RangiLyu) [inline .circleci/config.yml:79]: Add timm in `requirements/tests.txt` to avoid manually installing.

Comment (shinya7y) [reply .circleci/config.yml:79]: Should I add timm to `requirements/optional.txt` rather than `requirements/tests.txt`?

**Comment by ZwwWayne** [comment]
Can we directly use the TIMMBackbone in MMCls? I think we can directly use the config like

```
_base_ = [
    '../_base_/models/retinanet_r50_fpn.py',
    '../_base_/datasets/coco_detection.py',
    '../_base_/schedules/schedule_1x.py', '../_base_/default_runtime.py'
]
model = dict(
    backbone=dict(
        _delete_=True,
        type='mmcls.TIMMBackbone',
        model_name='resnet18',
        out_indices=(1, 2, 3, 4)),
    neck=dict(in_channels=[64, 128, 256, 512]))

# optimizer
optimizer = dict(type='SGD', lr=0.01, momentum=0.9, weight_decay=0.0001)
```
After we installed mmcls/timm

**Comment by shinya7y** [comment]
We need to update mmcls's TIMMBackbone for mmdet and mmseg because mmcls's TIMMBackbone does not support `features_only=True`.
After updating, the config will be as follows.

```python
_base_ = [
    '../_base_/models/retinanet_r50_fpn.py',
    '../_base_/datasets/coco_detection.py',
    '../_base_/schedules/schedule_1x.py', '../_base_/default_runtime.py'
]
model = dict(
    backbone=dict(
        _delete_=True,
        type='mmcls.TIMMBackbone',
        model_name='resnet18',
        features_only=True,  # mmcls assumes False only, causing error
        pretrained=True,
        out_indices=(1, 2, 3, 4)),
    neck=dict(in_channels=[64, 128, 256, 512]))

# optimizer
optimizer = dict(type='SGD', lr=0.01, momentum=0.9, weight_decay=0.0001)

# import to trigger register_module in mmcls
custom_imports = dict(imports=['mmcls.models'], allow_failed_imports=False)

# disable NumClassCheckHook
custom_hooks = []
```

**Comment by ZwwWayne** [comment]
> We need to update mmcls's TIMMBackbone for mmdet and mmseg because mmcls's TIMMBackbone does not support `features_only=True`. After updating, the config will be as follows.
> 
> ```python
> _base_ = [
>     '../_base_/models/retinanet_r50_fpn.py',
>     '../_base_/datasets/coco_detection.py',
>     '../_base_/schedules/schedule_1x.py', '../_base_/default_runtime.py'
> ]
> model = dict(
>     backbone=dict(
>         _delete_=True,
>         type='mmcls.TIMMBackbone',
>         model_name='resnet18',
>         features_only=True,  # mmcls assumes False only, causing error
>         pretrained=True,
>         out_indices=(1, 2, 3, 4)),
>     neck=dict(in_channels=[64, 128, 256, 512]))
> 
> # optimizer
> optimizer = dict(type='SGD', lr=0.01, momentum=0.9, weight_decay=0.0001)
> 
> # import to trigger register_module in mmcls
> custom_imports = dict(imports=['mmcls.models'], allow_failed_imports=False)
> 
> # disable NumClassCheckHook
> custom_hooks = []
> ```

Great! Thank you @shinya7y , would you like to directly create a PR in mmcls to support that? We can also create an issue there to let them fix this in this version.

File: .circleci/config.yml (hunk 1 above)
Comment (ZwwWayne) [reply .circleci/config.yml:79]: We can simply put it into requirements/optional.txt

**Comment by ZwwWayne** [comment]
Hi @shinya7y ,
It seems that mmcls already support that. We can add comments in the config to indicate that it requires the newest mmcls (should be v0.20.0) and clean the code. Hopefully this PR can be merged by this week as a feature in the next release of mmdet.

**Comment by shinya7y** [comment]
I will do them and put example configs in `configs/timm_example` like `configs/albu_example`.
If other directories (e.g., `configs/timm_backbone`, `docs/*/tutorials`) are preferred, please let me know.

**Comment by ZwwWayne** [comment]
> I will do them and put example configs in `configs/timm_example` like `configs/albu_example`. If other directories (e.g., `configs/timm_backbone`, `docs/*/tutorials`) are preferred, please let me know.

Hi @shinya7y,
This idea is great! I think we can do that~

**Comment by hhaAndroid** [comment]
@shinya7y Great.  LGTM.

**Comment by hhaAndroid** [comment]
@shinya7y The timm of mmcls cannot support frozen stages. Is there any way to do this?

**Comment by shinya7y** [comment]
> @shinya7y The timm of mmcls cannot support frozen stages. Is there any way to do this?

It will be difficult. Related discussion: https://github.com/open-mmlab/mmclassification/issues/605

**Comment by hhaAndroid** [comment]
> > @shinya7y The timm of mmcls cannot support frozen stages. Is there any way to do this?
> 
> It will be difficult. Related discussion: [open-mmlab/mmclassification#605](https://github.com/open-mmlab/mmclassification/issues/605)

OK.

**Review by hhaAndroid — APPROVED** [review]

**Review by RangiLyu — APPROVED** [review]
LGTM

**Review by ZwwWayne — APPROVED** [review]

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**